Mathlib/Algebra/Star/CentroidHom.lean

```diff
@@ -0,0 +1,136 @@
+Authors: Christopher Hoskin
+-/
+
+import Mathlib.Algebra.Ring.CentroidHom
+import Mathlib.Algebra.Star.StarRingHom
+import Mathlib.Algebra.Star.Subsemiring
+import Mathlib.Algebra.Star.Basic
+
+/-!
+# Centroid homomorphisms on Star Rings
+
+When a (non unital, non-associative) semi-ring is equipped with an involutive automorphism the
+center of the centroid becomes a star ring in a natural way and the natural mapping of the centre of
+the semi-ring into the centre of the centroid becomes a *-homomorphism.
+
+## Tags
+
+centroid
+-/
+
+variable {α : Type*}
+
+namespace CentroidHom
+
+section NonUnitalNonAssocStarSemiring
+
+variable [NonUnitalNonAssocSemiring α] [StarRing α]
+
+instance : Star (CentroidHom α) where
+  star f :=
+  { toFun := fun a => star (f (star a))
+    map_zero' := by
+      simp only [star_zero, map_zero]
+    map_add' := fun a b => by simp only [star_add, map_add]
+    map_mul_left' := fun a b => by simp only [star_mul, map_mul_right, star_star]
+    map_mul_right' := fun a b => by simp only [star_mul, map_mul_left, star_star] }
+
+@[simp] lemma star_apply (f : CentroidHom α) (a : α) : (star f) a = star (f (star a)) := rfl
+
+instance instStarAddMonoid : StarAddMonoid (CentroidHom α) where
+  star_involutive f := ext (fun _ => by
+    rw [star_apply, star_apply, star_star, star_star])
+  star_add f g := ext fun _ => star_add _ _
+
+instance : Star (Subsemiring.center (CentroidHom α)) where
+  star f := ⟨star (f : CentroidHom α), Subsemiring.mem_center_iff.mpr (fun g => ext (fun a =>
+    calc
+      g (star (f (star a))) = star (star g (f (star a))) := by rw [star_apply, star_star]
+      _ = star ((star g * f) (star a)) := rfl
+      _ = star ((f * star g) (star a)) := by rw [f.property.comm]
+      _ = star (f (star g (star a))) := rfl
+      _ = star (f (star (g a))) := by rw [star_apply, star_star]))⟩
+
+instance instStarAddMonoidCenter : StarAddMonoid (Subsemiring.center (CentroidHom α)) where
+  star_involutive f := SetCoe.ext (star_involutive f.val)
+  star_add f g := SetCoe.ext (star_add f.val g.val)
+
+instance : StarRing (Subsemiring.center (CentroidHom α)) where
+  __ := instStarAddMonoidCenter
+  star_mul f g := by
+    ext a
+    calc
+      star (f * g) a = star (g * f) a := by rw [CommMonoid.mul_comm f g]
+      _ = star (g (f (star a))) := rfl
+      _ = star (g (star (star (f (star a))))) := by simp only [star_star]
+      _ = (star g * star f) a := rfl
+
+/-- The canonical *-homomorphism embedding the center of `CentroidHom α` into `CentroidHom α`. -/
+def centerStarEmbedding : Subsemiring.center (CentroidHom α) →⋆ₙ+* CentroidHom α where
+  toNonUnitalRingHom :=
+    (SubsemiringClass.subtype (Subsemiring.center (CentroidHom α))).toNonUnitalRingHom
+  map_star' f := by
+    simp only [RingHom.toMonoidHom_eq_coe, OneHom.toFun_eq_coe, MonoidHom.toOneHom_coe,
+      MonoidHom.coe_coe, SubsemiringClass.coe_subtype]
+    exact rfl
+
+theorem inr_star (z : NonUnitalStarSubsemiring.center α) :
```

Comment:
What does `inr` refer to here? I think this should probably be called `star_centerToCentroidCenter`, or stated in reverse as `centerToCentroidCenter_star`.

Reply:
I can't remember now! Renamed.